File: 2° BackEnd/app/main/subModels/registry.py
from django.db import models
from main.subModels.game import Game
from main.subModels.user import User
from main.subModels.tag import Tag
from main.enum.gameEnum import GameEnum
from main.subModels.auth import Auth

PROGRESS_CHOICES = [
    (GameEnum.PLAN.name, GameEnum.PLAN.value), 
    (GameEnum.PLAY.name, GameEnum.PLAY.value),
    (GameEnum.DROP.name, GameEnum.DROP.value),
    (GameEnum.HOLD.name, GameEnum.HOLD.value),
    (GameEnum.REPLAY.name, GameEnum.REPLAY.value),
    (GameEnum.COMPLETE.name, GameEnum.COMPLETE.value)
]

NOTE_CHOICES = [
    ('3', GameEnum.HORRIBLE.value),
    ('5', GameEnum.MEH.value),
    ('7', GameEnum.FINE.value),
    ('8', GameEnum.GOOD.value),
    ('9', GameEnum.GREAT.value),
    ('10', GameEnum.BEST.value)
]

class Registry(models.Model):
    progress = models.CharField(max_length = 30, default=GameEnum.PLAN.name, choices=PROGRESS_CHOICES)
    note = models.CharField(max_length = 30, null=True, default=None, choices=NOTE_CHOICES)
    favorite = models.BooleanField(default=False)
    updateDate = models.DateTimeField(auto_now=True)
    recommendation = models.BooleanField(default=False)
    game = models.ForeignKey(Game, on_delete=models.CASCADE)
    user = models.ForeignKey(User, on_delete=models.CASCADE)
    tag = models.ForeignKey(Tag, on_delete=models.CASCADE)

    class Meta:
        constraints = [
            models.UniqueConstraint(fields=['user', 'game'], name='unique_user_game')
        ]

    def __str__(self):
        return f"\n| Id: {self.id}\n| Note: {self.note}\n UpdateDate: {self.updateDate}\n Favorite: {self.favorite}\n Progress: {self.progress} |\n"

File: 3° FrontEnd/src/app/shared/models/Registry.ts
import { Game } from "./Game"
import { Tag } from "./Tag"
import { User } from "./User"

export class Registry {
    private _id?: number | undefined
    private _progress?: string
    private _note?: string
    private _updateDate?: string
    private _favorite?: boolean
    private _recommendation?: boolean
    private _tag!: Tag
    private _user!: User
    private _game!: Game

    constructor(id: number, progress: string, note: string | undefined, updateDate: string | undefined, favorite: boolean, recommendation: boolean, tag: Tag, user: User, game: Game);
    constructor(progress: string, note: string  | undefined, favorite: string, recommendation: boolean, tag: Tag, user: User, game: Game);
    constructor(user: User, game: Game, tag: Tag);
    constructor();
    constructor(...myarray: any[]) {
        if (myarray.length === 9) {
            this._id = myarray[0]
            this._progress = myarray[1]
            this._note = myarray[2]
            this._updateDate = myarray[3]
            this._favorite = myarray[4]
            this._recommendation = myarray[5]
            this._tag = myarray[6]
            this._user = myarray[7]
            this._game = myarray[8]
            return;
        }
        if (myarray.length === 7) {
            this._progress = myarray[0]
            this._note = myarray[1]
            this._favorite = myarray[2]
            this._recommendation = myarray[3]
            this._tag = myarray[4]
            this._user = myarray[5]
            this._game = myarray[6]
            return;
        }
        if (myarray.length === 3) {
            this._user = myarray[0]
            this._game = myarray[1]
            this._tag = myarray[2]
            return;
        }
    }

    static simpleRefract(registry: Registry) {
        const result = {
            id: registry!.id,
            progress: registry!.progress,
            note: registry!.note,
            updateDate: registry!.updateDate,
            favorite: registry!.favorite,
            recommendation: registry!.recommendation,
            tag: registry!.tag.id,
            user: registry!.user.id,
            game: registry!.game.id
        }
        return result;
    }

    public get id(): number | undefined {
        return this._id
    }
    public set id(value: number | undefined) {
        this._id = value
    }

    public get progress(): string | undefined {
        return this._progress
    }
    public set progress(value: string | undefined) {
        this._progress = value
    }

    public get favorite(): boolean | undefined {
        return this._favorite
    }
    public set favorite(value: boolean | undefined) {
        this._favorite = value
    }

    public get recommendation(): boolean | undefined {
        return this._recommendation
    }
    public set recommendation(value: boolean | undefined) {
        this._recommendation = value
    }

    get note(): string | undefined {
        return this._note
    }
    set note(value: string | undefined) {
        this._note = value
    }

    get updateDate(): string | undefined {
        return this._updateDate
    }
    set updateDate(value: string | undefined) {
        this._updateDate = value
    }

    get tag(): Tag {
        return this._tag
    }

    set tag(value: Tag) {
        this._tag = value
    }

    public get user(): User {
        return this._user
    }
    public set user(value: User) {
        this._user = value
    }

    public get game(): Game {
        return this._game
    }
    public set game(value: Game) {
        this._game = value
    }
}